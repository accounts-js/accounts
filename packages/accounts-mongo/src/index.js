// @flow

import { get } from 'lodash';
import { encryption } from '@accounts/server';
import type {
  CreateUserType,
  UserObjectType,
  SessionType,
} from '@accounts/common';
import { ObjectID } from 'mongodb';

export type MongoOptionsType = {
  collectionName: string,
  sessionCollectionName: string,
  timestamps: {
    createdAt: string,
    updatedAt: string,
  }
};

export type MongoUserObjectType = {
  username?: string,
  profile?: Object,
  services: {
    password?: {
      bcrypt: string,
    },
  },
  emails?: [{
    address: string,
    verified: boolean,
  }],
};

const toMongoID = (objectId) => {
  if (typeof objectId === 'string') {
    return new ObjectID(objectId);
  }

  return objectId;
};

class Mongo {
  options: MongoOptionsType;
  // TODO definition for mongodb connection object
  db: any;
  collection: any;
  sessionCollection: any;

  constructor(db: any, options: MongoOptionsType) {
    const defaultOptions = {
      collectionName: 'users',
      sessionCollectionName: 'sessions',
      timestamps: {
        createdAt: 'createdAt',
        updatedAt: 'updatedAt',
      },
    };
    this.options = { ...defaultOptions, ...options };
    if (get(db, 'constructor.define.name') !== 'Db') {
      throw new Error('A valid database connection object is required');
    }
    this.db = db;
    this.collection = this.db.collection(this.options.collectionName);
    this.sessionCollection = this.db.collection(this.options.sessionCollectionName);
  }

  async setupIndexes(): Promise<void> {
    await this.collection.createIndex('username', { unique: 1, sparse: 1 });
    await this.collection.createIndex('emails.address', { unique: 1, sparse: 1 });
  }

  async createUser(options: CreateUserType): Promise<string> {
    const user: MongoUserObjectType = {
      services: {},
      profile: {},
      [this.options.timestamps.createdAt]: Date.now(),
      [this.options.timestamps.updatedAt]: Date.now(),
    };
    if (options.password) {
      user.services.password = { bcrypt: await encryption.hashPassword(options.password) };
    }
    if (options.username) {
      user.username = options.username;
    }
    if (options.email) {
      user.emails = [{ address: options.email.toLowerCase(), verified: false }];
    }
    if (options.profile) {
      user.profile = options.profile;
    }
    const ret = await this.collection.insertOne(user);
    return ret.ops[0]._id;
  }

  async findUserById(userId: string): Promise<?UserObjectType> {
    const user = await this.collection.findOne({ _id: toMongoID(userId) });
    if (user) {
      user.id = user._id;
    }
    return user;
  }

  async findUserByEmail(email: string): Promise<?UserObjectType> {
    const user = await this.collection.findOne({ 'emails.address': email.toLowerCase() });
    if (user) {
      user.id = user._id;
    }
    return user;
  }

  async findUserByUsername(username: string): Promise<?UserObjectType> {
    const user = await this.collection.findOne({ username });
    if (user) {
      user.id = user._id;
    }
    return user;
  }

  async findPasswordHash(userId: string): Promise<?string> {
    const user = await this.findUserById(toMongoID(userId));
    if (user) {
      return user.services.password.bcrypt;
    }
    return null;
  }

  async addEmail(userId: string, newEmail: string, verified: boolean): Promise<void> {
    const ret = await this.collection.update({ _id: toMongoID(userId) }, {
      $addToSet: {
        emails: {
          address: newEmail.toLowerCase(),
          verified,
        },
      },
      $set: { [this.options.timestamps.updatedAt]: Date.now() },
    });
    if (ret.result.nModified === 0) {
      throw new Error('User not found');
    }
  }

  async removeEmail(userId: string, email: string): Promise<void> {
    const ret = await this.collection.update({ _id: toMongoID(userId) }, {
      $pull: { emails: { address: email.toLowerCase() } },
      $set: { [this.options.timestamps.updatedAt]: Date.now() },
    });
    if (ret.result.nModified === 0) {
      throw new Error('User not found');
    }
  }

  async setUsername(userId: string, newUsername: string): Promise<void> {
    const ret = await this.collection.update({ _id: toMongoID(userId) }, {
      $set: {
        username: newUsername,
        [this.options.timestamps.updatedAt]: Date.now(),
      },
    });
    if (ret.result.nModified === 0) {
      throw new Error('User not found');
    }
  }

  async setPasssword(userId: string, newPassword: string): Promise<void> {
    const ret = await this.collection.update({ _id: toMongoID(userId) }, {
      $set: {
        'services.password.bcrypt': await encryption.hashPassword(newPassword),
        [this.options.timestamps.updatedAt]: Date.now(),
      },
    });
    if (ret.result.nModified === 0) {
      throw new Error('User not found');
    }
  }

  async setProfile(userId: string, profile: Object): Promise<Object> {
    await this.collection.update({ _id: toMongoID(userId) }, {
      $set: {
        profile,
        [this.options.timestamps.updatedAt]: Date.now(),
      },
    });
    const user = await this.findUserById(userId);
    return user && user.profile;
  }

  async createSession(userId: string, ip: string, userAgent: string): Promise<string> {
    const ret = await this.sessionCollection.insertOne({
      userId,
      userAgent,
      ip,
      valid: true,
      [this.options.timestamps.createdAt]: Date.now(),
      [this.options.timestamps.updatedAt]: Date.now(),
    });
    return ret.ops[0]._id;
  }

  async updateSession(sessionId: string, ip: string, userAgent: string): Promise<void> {
    await this.sessionCollection.update({ _id: toMongoID(sessionId) }, {
      $set: {
        ip,
        userAgent,
        [this.options.timestamps.updatedAt]: Date.now(),
      },
    });
  }

  async invalidateSession(sessionId: string): Promise<void> {
    await this.sessionCollection.update({ _id: toMongoID(sessionId) }, {
      $set: {
        valid: false,
        [this.options.timestamps.updatedAt]: Date.now(),
      },
    });
  }

  findSessionById(sessionId: string): Promise<?SessionType> {
    return this.sessionCollection.findOne({ _id: toMongoID(sessionId) });
  }
}

export default Mongo;
