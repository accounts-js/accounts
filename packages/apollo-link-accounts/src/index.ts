import { setContext } from 'apollo-link-context';
import { ApolloLink } from 'apollo-link';
import { AccountsClient } from '@accounts/client';

export const accountsLink = (accountsClient: AccountsClient): ApolloLink => {
  return setContext(async (_, { headers }) => {
    const tokens = await accountsClient.refreshSession();
    return {
      headers: {
        ...headers,
        'accounts-access-token': tokens ? tokens.accessToken : null,
      },
    };
  });
};
