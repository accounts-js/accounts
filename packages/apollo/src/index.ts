import { accountsLink } from '@accounts/apollo-link';
import { AccountsClientOptions, AccountsClient } from '@accounts/client';
import AccountsGraphQLClient from '@accounts/graphql-client';
import { ApolloClient } from 'apollo-client';
import { ApolloLink } from 'apollo-link';
import { onError } from 'apollo-link-error';
import { HttpLink } from 'apollo-link-http';
import { InMemoryCache } from 'apollo-cache-inmemory';

export interface AccountsApolloOptions extends AccountsClientOptions {
  uri: string;
}

export const accountsApollo = (options: AccountsApolloOptions) => {
  const accountsApolloClient = new ApolloClient({
    cache: new InMemoryCache(),
    link: ApolloLink.from([
      onError(({ graphQLErrors, networkError }) => {
        if (graphQLErrors) {
          graphQLErrors.map(({ message, locations, path }) =>
            console.log(
              `[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`
            )
          );
        }
        if (networkError) {
          console.log(`[Network error]: ${networkError}`);
        }
      }),
      new HttpLink({
        uri: options.uri,
      }),
    ]),
  });

  const accountsClient = new AccountsClient(
    options,
    new AccountsGraphQLClient({
      graphQLClient: accountsApolloClient,
    })
  );

  let accountsPassword;
  if (require.resolve('@accounts/client-password')) {
    const AccountsClientPassword = require('@accounts/client-password').AccountsClientPassword; // tslint:disable-line no-implicit-dependencies
    accountsPassword = new AccountsClientPassword(accountsClient);
  }

  return {
    accountsClient,
    accountsApolloClient,
    accountsLink: accountsLink(accountsClient),
    accountsPassword,
  };
};

export default accountsApollo;
