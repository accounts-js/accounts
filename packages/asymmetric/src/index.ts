import AccountsAsymmetric from './accounts-asymmetric';

export default AccountsAsymmetric;
export { AccountsAsymmetric };
