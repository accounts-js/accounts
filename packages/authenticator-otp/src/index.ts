import {
  DatabaseInterface,
  AuthenticatorService,
  Authenticator,
  MfaChallenge,
} from '@accounts/types';
import { AccountsServer, generateRandomToken } from '@accounts/server';
import { authenticator as optlibAuthenticator } from 'otplib';

// TODO remove the url to be returned
// - should be done client side
// - update documentation to see how to use it

interface DbAuthenticatorOtp extends Authenticator {
  secret: string;
}

export interface AuthenticatorOtpOptions {
  /**
   * Two factor app name that will be displayed inside the user authenticator app.
   */
  appName?: string;

  /**
   * Two factor user name that will be displayed inside the user authenticator app,
   * usually a name, email etc..
   * Will be called every time a user register a new device.
   * That way you can display something like "Github (leo@accountsjs.com)" in the authenticator app.
   */
  userName?: (userId: string) => Promise<string> | string;
}

const defaultOptions = {
  appName: 'accounts-js',
};

export class AuthenticatorOtp implements AuthenticatorService {
  public authenticatorName = 'otp';
  public server!: AccountsServer;

  private options: AuthenticatorOtpOptions & typeof defaultOptions;
  private db!: DatabaseInterface;

  constructor(options: AuthenticatorOtpOptions = {}) {
    this.options = { ...defaultOptions, ...options };
  }

  public setStore(store: DatabaseInterface) {
    this.db = store;
  }

  /**
   * @description Start the association of a new OTP device
   */
  public async associate(
    userIdOrMfaChallenge: string | MfaChallenge
  ): Promise<{ id: string; mfaToken: string; secret: string; otpauthUri: string }> {
    const userId =
      typeof userIdOrMfaChallenge === 'string' ? userIdOrMfaChallenge : userIdOrMfaChallenge.userId;
    const mfaChallenge = typeof userIdOrMfaChallenge === 'string' ? null : userIdOrMfaChallenge;

    const secret = optlibAuthenticator.generateSecret();
    const userName = this.options.userName ? await this.options.userName(userId) : userId;
    const otpauthUri = optlibAuthenticator.keyuri(userName, this.options.appName, secret);

    const authenticatorId = await this.db.createAuthenticator({
      type: this.authenticatorName,
      userId,
      secret,
      active: false,
    });

    let mfaToken: string;
    if (mfaChallenge) {
      mfaToken = mfaChallenge.token;
      await this.db.updateMfaChallenge(mfaChallenge.id, {
        authenticatorId,
      });
    } else {
      // We create a new challenge for the authenticator so it can be verified later
      mfaToken = generateRandomToken();
      await this.db.createMfaChallenge({
        userId,
        authenticatorId,
        token: mfaToken,
        scope: 'associate',
      });
    }

    return {
      id: authenticatorId,
      mfaToken,
      secret,
      otpauthUri,
    };
  }

  /**
   * @description Verify that the code provided by the user is valid
   */
  public async authenticate(
    mfaChallenge: MfaChallenge,
    authenticator: DbAuthenticatorOtp,
    { code }: { code?: string }
  ): Promise<boolean> {
    if (!code) {
      throw new Error('Code required');
    }

    return optlibAuthenticator.check(code, authenticator.secret);
  }

  /**
   * @description Remove the sensitive fields from the database authenticator. The object
   * returned by this function can be exposed to the user safely.
   */
  public sanitize(authenticator: DbAuthenticatorOtp): Authenticator {
    // The secret key should never be exposed to the user after the authenticator is linked
    const {
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      secret,
      ...safeAuthenticator
    } = authenticator;
    return safeAuthenticator;
  }
}
