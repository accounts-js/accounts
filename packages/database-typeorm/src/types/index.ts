import { User } from '../entity/User';
import { UserEmail } from '../entity/UserEmail';
import { UserService } from '../entity/UserService';
import { UserSession } from '../entity/UserSession';

export interface AccountsTypeormOptions {
  connectionName?: string;
  userEntity?: typeof User;
  userServiceEntity?: typeof UserService;
  userEmailEntity?: typeof UserEmail;
  userSessionEntity?: typeof UserSession;
}
