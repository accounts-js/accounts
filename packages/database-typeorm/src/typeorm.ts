import { ConnectionInformations, CreateUser, DatabaseInterface } from '@accounts/types';
import { Repository, getRepository, Not, In, FindOperator } from 'typeorm';
import { User } from './entity/User';
import { UserEmail } from './entity/UserEmail';
import { UserService } from './entity/UserService';
import { UserSession } from './entity/UserSession';
import { AccountsTypeormOptions } from './types';

const defaultOptions = {
  userEntity: User,
  userEmailEntity: UserEmail,
  userServiceEntity: UserService,
  userSessionEntity: UserSession,
};

export class AccountsTypeorm implements DatabaseInterface {
  private options: AccountsTypeormOptions & typeof defaultOptions;
  private userRepository: Repository<User> = null as any;
  private emailRepository: Repository<UserEmail> = null as any;
  private serviceRepository: Repository<UserService> = null as any;
  private sessionRepository: Repository<UserSession> = null as any;

  constructor(options?: AccountsTypeormOptions) {
    this.options = { ...defaultOptions, ...options };

    const {
      connection,
      connectionName,
      userEntity,
      userEmailEntity,
      userServiceEntity,
      userSessionEntity,
    } = this.options;

    const setRepositories = () => {
      if (connection) {
        this.userRepository = connection.getRepository(userEntity);
        this.emailRepository = connection.getRepository(userEmailEntity);
        this.serviceRepository = connection.getRepository(userServiceEntity);
        this.sessionRepository = connection.getRepository(userSessionEntity);
      } else {
        this.userRepository = getRepository(userEntity, connectionName);
        this.emailRepository = getRepository(userEmailEntity, connectionName);
        this.serviceRepository = getRepository(userServiceEntity, connectionName);
        this.sessionRepository = getRepository(userSessionEntity, connectionName);
      }
    };

    // direct or lazy support
    if (connection && !connection.isConnected) {
      connection.connect().then(setRepositories);
    } else {
      setRepositories();
    }
  }

  public async findUserByEmail(email: string): Promise<User | null> {
    const userEmail = await this.emailRepository.findOne({
      where: { address: email.toLocaleLowerCase() },
      cache: this.options.cache,
    });

    if (userEmail) {
      return this.findUserById(userEmail.userId);
    }

    return null;
  }

  public async findUserByUsername(username: string): Promise<User | null> {
    const user = await this.userRepository.findOne({
      where: { username },
      cache: this.options.cache,
    });

    if (user) {
      return user;
    }

    return null;
  }

  public async findUserById(userId: string): Promise<User | null> {
    const user = await this.userRepository.findOne(userId, {
      cache: this.options.cache,
    });
    if (!user) {
      // throw new Error('User not found');
      return null;
    }
    return user;
  }

  public async findUserByResetPasswordToken(token: string): Promise<User | null> {
    const service = await this.serviceRepository.findOne({
      where: {
        name: 'password.reset',
        token,
      },
      cache: this.options.cache,
    });

    if (service) {
      return this.findUserById(service.userId);
    }

    return null;
  }

  public async findUserByEmailVerificationToken(token: string): Promise<User | null> {
    const service = await this.serviceRepository.findOne({
      where: {
        name: 'email.verificationTokens',
        token,
      },
      cache: this.options.cache,
    });

    if (service) {
      return this.findUserById(service.userId);
    }

    return null;
  }

  public async createUser(createUser: CreateUser): Promise<string> {
    const { username, email, password, ...otherFields } = createUser;

    const user = new this.options.userEntity();

    if (email) {
      const userEmail = new this.options.userEmailEntity();
      userEmail.address = email.toLocaleLowerCase();
      userEmail.verified = false;
      await this.emailRepository.save(userEmail);
      user.emails = [userEmail];
    }

    if (password) {
      const userService = new this.options.userServiceEntity();
      userService.name = 'password';
      userService.options = { bcrypt: password };
      await this.serviceRepository.save(userService);
      user.allServices = [userService];
    }

    if (username) {
      user.username = username;
    }

    Object.assign(user, otherFields);

    await this.userRepository.save(user);

    return user.id;
  }

  public async setUsername(userId: string, newUsername: string): Promise<void> {
    const user = await this.findUserById(userId);
    if (user) {
      user.username = newUsername;
      await this.userRepository.save(user);
      return;
    }
    throw new Error('User not found');
  }

  public async findUserByServiceId(serviceName: string, serviceId: string): Promise<User | null> {
    const service = await this.serviceRepository.findOne({
      where: {
        name: serviceName,
        serviceId,
      },
      cache: this.options.cache,
    });

    if (service) {
      return this.findUserById(service.userId);
    }

    return null;
  }

  public async getService(userId: string, serviceName: string): Promise<UserService | null> {
    const user = await this.findUserById(userId);
    if (user) {
      const service = user.allServices.find((s) => s.name === serviceName);
      if (service) {
        return service;
      }
    }

    return null;
  }

  public async setService(
    userId: string,
    serviceName: string,
    data: object,
    token?: string
  ): Promise<void> {
    let service = await this.getService(userId, serviceName);

    if (!service) {
      const user = await this.findUserById(userId);
      if (user) {
        service = new this.options.userServiceEntity();
        service.name = serviceName;
        service.user = user;
      }
    }

    const { id = null, ...options } = data as any;

    if (service) {
      service.options = options;

      if (id) {
        service.serviceId = id;
      }

      if (token) {
        service.token = token;
      }

      await this.serviceRepository.save(service);
    }
  }

  public async unsetService(userId: string, serviceName: string): Promise<void> {
    const user = await this.findUserById(userId);
    if (user) {
      const service = user.allServices.find((s) => s.name === serviceName);
      if (service) {
        await this.serviceRepository.remove(service);
      }
    }
  }

  public async findPasswordHash(userId: string): Promise<string | null> {
    const service = await this.getService(userId, 'password');
    if (service) {
      return service.options.bcrypt;
    }
    return null;
  }

  public async setPassword(userId: string, newPassword: string): Promise<void> {
    const user = await this.findUserById(userId);
    if (user) {
      await this.setService(userId, 'password', { bcrypt: newPassword });
      await this.userRepository.update({ id: user.id }, {});
      return;
    }
    throw new Error('User not found');
  }

  public async addResetPasswordToken(
    userId: string,
    email: string,
    token: string,
    reason: string
  ): Promise<void> {
    await this.setService(
      userId,
      'password.reset',
      {
        address: email.toLocaleLowerCase(),
        when: new Date().toJSON(),
        reason,
      },
      token
    );
  }

  public async addEmail(userId: string, newEmail: string, verified: boolean): Promise<void> {
    const user = await this.findUserById(userId);
    if (user) {
      const userEmail = new this.options.userEmailEntity();
      userEmail.user = user;
      userEmail.address = newEmail.toLocaleLowerCase();
      userEmail.verified = verified;
      await this.emailRepository.save(userEmail);
      await this.userRepository.update({ id: user.id }, {});
      return;
    }
    throw new Error('User not found');
  }

  public async removeEmail(userId: string, email: string): Promise<void> {
    const user = await this.findUserById(userId);
    if (user) {
      const userEmail = user.emails.find((s) => s.address === email.toLocaleLowerCase());
      if (!userEmail) {
        throw new Error('Email not found');
      }
      await this.emailRepository.remove(userEmail);
      await this.userRepository.update({ id: user.id }, {});
      return;
    }
    throw new Error('User not found');
  }

  public async verifyEmail(userId: string, email: string): Promise<void> {
    const user = await this.findUserById(userId);
    if (user) {
      const userEmail = user.emails.find((s) => s.address === email.toLocaleLowerCase());

      if (!userEmail) {
        throw new Error('Email not found');
      }
      userEmail.verified = true;
      await this.emailRepository.save(userEmail);
      await this.unsetService(userId, 'email.verificationTokens');
      await this.userRepository.update({ id: user.id }, {});
      return;
    }
    throw new Error('User not found');
  }

  public async addEmailVerificationToken(
    userId: string,
    email: string,
    token: string
  ): Promise<void> {
    await this.setService(
      userId,
      'email.verificationTokens',
      {
        address: email.toLocaleLowerCase(),
        when: new Date(),
      },
      token
    );
  }

  public async removeAllResetPasswordTokens(userId: string): Promise<void> {
    await this.unsetService(userId, 'password.reset');
  }

  public async setUserDeactivated(userId: string, deactivated: boolean): Promise<void> {
    const user = await this.findUserById(userId);
    if (user) {
      user.deactivated = deactivated;
      await this.userRepository.save(user);
    }
  }

  public async findSessionById(sessionId: string): Promise<UserSession | null> {
    try {
      const session = await this.sessionRepository.findOne(sessionId, {
        cache: this.options.cache,
      });

      if (session) {
        return session;
      }
    } catch (err) {
      // noop
    }

    return null;
  }

  public async findSessionByToken(token: string) {
    const session = await this.sessionRepository.findOne(
      { token },
      {
        cache: this.options.cache,
      }
    );
    if (!session) {
      return null;
    }
    return session;
  }

  public async findUserByLoginToken(token: string): Promise<User | null> {
    const service = await this.serviceRepository.findOne({
      where: {
        name: 'magicLink.loginTokens',
        token,
      },
      cache: this.options.cache,
    });

    if (service) {
      return this.findUserById(service.userId);
    }

    return null;
  }

  public async addLoginToken(userId: string, email: string, token: string) {
    await this.setService(
      userId,
      'magicLink.loginTokens',
      {
        address: email.toLocaleLowerCase(),
        when: new Date().toJSON(),
      },
      token
    );
  }

  public async removeAllLoginTokens(userId: string) {
    await this.unsetService(userId, 'magicLink.loginTokens');
  }

  public async createSession(
    userId: string,
    token: string,
    connection: ConnectionInformations = {},
    extra?: object
  ) {
    const user = await this.findUserById(userId);
    const session = new this.options.userSessionEntity();
    session.user = user!;
    session.token = token;
    session.userAgent = connection.userAgent;
    session.ip = connection.ip;
    if (extra) {
      session.extra = extra;
    }
    session.valid = true;
    await this.sessionRepository.save(session);

    return session.id;
  }

  public async updateSession(sessionId: string, connection: ConnectionInformations): Promise<void> {
    const session = await this.findSessionById(sessionId);
    if (session) {
      session.userAgent = connection.userAgent;
      session.ip = connection.ip;
      await this.sessionRepository.save(session);
    }
  }

  public async invalidateSession(sessionId: string): Promise<void> {
    const session = await this.findSessionById(sessionId);
    if (session) {
      session.valid = false;
      await this.sessionRepository.save(session);
    }
  }

  public async invalidateAllSessions(userId: string, excludedSessionIds?: string[]): Promise<void> {
    const selector: { userId: string; id?: FindOperator<any> } = { userId };

    if (excludedSessionIds && excludedSessionIds.length > 0) {
      selector.id = Not(In(excludedSessionIds));
    }

    await this.sessionRepository.update(selector, {
      valid: false,
    });
  }
}
