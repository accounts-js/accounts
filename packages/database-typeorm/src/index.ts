import { AccountsTypeorm } from './typeorm';

export { AccountsTypeorm };
export default AccountsTypeorm;
