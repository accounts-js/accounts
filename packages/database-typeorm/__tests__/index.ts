import 'reflect-metadata';
import { Typeorm } from '../src/typeorm';

describe('Typeorm', () => {
  it('should have named export Typeorm', () => {
    expect(typeof Typeorm).toBe('function');
  });
});
