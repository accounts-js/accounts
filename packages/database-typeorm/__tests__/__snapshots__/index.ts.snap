// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountsTypeorm createUser should create a new user 1`] = `
Object {
  "address": "foo@bar.baz",
  "id": Any<String>,
  "userId": Any<String>,
  "verified": false,
}
`;

exports[`AccountsTypeorm createUser should create a new user 2`] = `
Object {
  "password": Array [
    Object {
      "bcrypt": "foobarbaz",
    },
  ],
}
`;
